controller/qname.go

```diff
@@ -0,0 +1,75 @@
+package controller
+
+import (
+	"fmt"
+	"strings"
+)
+
+// QName is a string type that represents a fully qualified name that includes the
+// partition, namespace and service name.
+type QName struct {
+	partition string
+	namespace string
+	name      string
+}
+
+// NewQName creates a QName from the given partition, namespace and name.
+// The partition and namespace are optional and can be set to the empty string
+// if they are not applicable.
+func NewQName(partition, namespace, name string) QName {
+	return QName{partition: partition, namespace: namespace, name: name}
+}
+
+// QNameFromString creates a QName from the given slash-separated string.
+func QNameFromString(s string) QName {
+	var partition, namespace, name string
+
+	parts := strings.SplitN(s, "/", 3)
+	if len(parts) > 2 {
+		partition = parts[0]
+		namespace = parts[1]
+		name = parts[2]
+	} else {
+		name = parts[len(parts)-1]
+	}
+
+	return QName{partition: partition, namespace: namespace, name: name}
+}
+
+// Partition returns the partition from the qualified name.
+// It returns the empty string if the QName does not contain
+// a partition and namespace.
+func (q QName) Partition() string {
+	if q.partition != "" && q.namespace != "" {
```

Comment:
Yes, I agree with using a data struct for the `ServiceName`. I would like to discuss the UX of the task configuration and how we want to handle the different permutations of the presence (or absence) of the tags partition/namespace tags.

Reply:
@erichaberkorn @pglass Please confirm that per our meeting the following is the behaviour we will support for the tags on service tasks:

| Task Tag | ACL Controller Behaviour |
| - | - |
| `consul.hashicorp.com/mesh` | Indicates that the task is part of the service mesh. If this tag is not present the task is ignored by the ACL Controller. |
| `consul.hashicorp.com/service-name` | Name of the service. This tag is optional and the value is used if present; otherwise, the service name is inferred from task ARN. |
| `consul.hashicorp.com/partition` | `Enterprise` Admin partition the service belongs to. This tag is optional and if it is not present the `default` partition is used. If it is present the `consul.hashicorp.com/namespace` tag must also be included in the task definition or the ACL controller will log an error and ignore the task. |
| `consul.hashicorp.com/namespace` | `Enterprise` Namespace the service belongs to within the partition. This tag is optional and if it is not present the `default` namespace is used. If it is present the `consul.hashicorp.com/partition` tag must also be included in the task definition or the ACL controller will log an error and ignore the task. |